spydra/src/main/java/com/spotify/spydra/submitter/api/ClusterPlacement.java

```diff
@@ -0,0 +1,94 @@
+package com.spotify.spydra.submitter.api;
+
+
+import static com.spotify.spydra.submitter.api.PoolingSubmitter.SPYDRA_PLACEMENT_TOKEN_LABEL;
+import static com.spotify.spydra.submitter.api.PoolingSubmitter.SPYDRA_UNPLACED_TOKEN;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+import com.spotify.spydra.api.model.Cluster;
+import com.spotify.spydra.model.SpydraArgument;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.IntStream;
+import org.immutables.value.Value;
+
+@Value.Style(
+    visibility = Value.Style.ImplementationVisibility.PRIVATE,
+    builderVisibility = Value.Style.BuilderVisibility.PACKAGE,
+    overshadowImplementation = true)
+@Value.Immutable
+abstract class ClusterPlacement {
+
+  abstract int clusterNumber();
+
+  abstract long clusterGeneration();
+
+  String token() {
+    return String.format("%s-%s", clusterNumber(), clusterGeneration());
+  }
+
+  static ClusterPlacement from(String token) {
+    String[] splits = token.split("-");
+    if (splits.length != 2) {
+      throw new IllegalArgumentException("Could not parse token: " + token);
+    }
+    try {
+      return new ClusterPlacementBuilder()
+          .clusterNumber(Integer.valueOf(splits[0]))
+          .clusterGeneration(Long.valueOf(splits[1]))
+          .build();
+    } catch (NumberFormatException nfe) {
+      throw new IllegalArgumentException("Could not parse token: " + token, nfe);
+    }
+  }
+
+  static List<ClusterPlacement> all(Supplier<Long> timeSource, SpydraArgument.Pooling pooling) {
+    return IntStream.range(0, pooling.getLimit())
+        .mapToObj(clusterNumber -> createClusterPlacement(timeSource, clusterNumber, pooling))
+        .collect(toList());
+  }
+
+  static ClusterPlacement createClusterPlacement(Supplier<Long> timeSource,
+                                                 int clusterNumber,
+                                                 SpydraArgument.Pooling pooling) {
+    int limit = pooling.getLimit();
+    long time = timeSource.get() / 1000;
+    long age = pooling.getMaxAge().getSeconds();
+
+    long generation = computeGeneration(clusterNumber, limit, time, age);
+
+    return new ClusterPlacementBuilder()
+        .clusterNumber(clusterNumber)
+        .clusterGeneration(generation)
+        .build();
+  }
+
+  static long computeGeneration(int clusterNumber, int limit, long time, long age) {
```

Comment:
I think we can simply fulfill the case of scaling up/down the cluster (user increase decrease  pool limit in args for a newly submitted job), and it can be done by making formula independent of limit, for instance clusterNumber + (time/age % 2 - then we get 2 interleaving cluster for each clusternumber to deal with retiring cluster - we might need to rethink the name generation then :).
In that case you can increase or decrease pool limit, given we use age for retiring clusters

Reply:
The limit is used to add an offset so that the life of all the clusters does not start and end at the same time

Reply:
life of all clusters staring and ending at the same time is not a concern as we use gonna use idle time to terminate a cluster. I dont see any practical downside

Reply:
on the other hand getting rid limit here will give us easy up/down scaling of pool

Reply:
There is no guarantee for clusters to be idle, especially in the case of pooling it is very likely (and desired!) that jobs are uniformly spread.

Reply:
Aside, changing the parameters takes just one period to have effect and become eventually consistent, I don’t see this as a concern for scaling the number of clusters.

Reply:
clusters are guaranteed to become idle as at some point in time we submit job requests to new generations of clusters. and in case of changing limit parameters, if its changes several time within a single period (length of age) we can get large number of clusters, if we assume it happens just once for in a long time (which is most likely scenario but not worst case), then yes what you said is correct!

Reply:
@kiarash-rezahanjani I don't see how removing the limit will improve the scaling of the number of clusters. It will be the same but we will loose the offset, which is useful in case most of the jobs arrive at the same time, so some of them can be scheduled into clusters that are already running instead of having to wait until all of them are created